### PR TITLE
Add flex version of evaluateDerivatives

### PR DESCRIPTION
**Pull request by markdewing**
## Proposed changes

Make it clear that this function evaluates derivatives of the variational parameters (vs. spatial derivatives) by adding 'Parameter' to the name.

This only implements the first level of the flex_* routine.  Deeper levels need to be implemented to run in nested parallel or with offload.  However, this will allow the optimizer code to change.
    
The original version works on one configuration and returns an array with one element per variational parameter (well, two arrays).   The flex version will evaluate several configurations, and needs to return an array of arrays.  Possible solutions are to use a Matrix, use nested std::vector, or create a new data structure.   We use the last approach by creating the RecordArray class.  It should provide some abstraction between the structure and the storage used.

There is an incidental change for the unit tests to make open boundary conditions work in complex builds. The unit test uses atomic orbitals, and fails with the complex build in EvalPeriodicImagePhaseFactors with the "Attempting to run PBC LCAO with no HDF5 support" error.
For actual runs, the complex build should not be used with LCAO and open boundary conditions (??).  If we want to enforce this, it would be nice to have an error message stating that.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
@anbenali Please can you study the state of LCAO with complex and create issues as needed?. There are previous reports of breakage. However I am not sure if this is new breakage or a gap in implementation.

File: src/Containers/MinimalContainers/RecordArray.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+//
+// File developed by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//
+// File created by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef QMCPLUSPLUS_RECORDARRAY_H
+#define QMCPLUSPLUS_RECORDARRAY_H
+
+#include <stdio.h>
```
Comment (ye-luo) [inline src/Containers/MinimalContainers/RecordArray.hpp:15]: Is this really needed? If yes, ue C++ header <cstdio> is better.

File: src/QMCWaveFunctions/TrialWaveFunction.h [hunk 2]
```diff
@@ -318,6 +319,12 @@ class TrialWaveFunction : public MPIObjectBase
                            std::vector<ValueType>& dhpsioverpsi,
                            bool project = false);
 
+  void flex_evaluateParameterDerivatives(const RefVector<TrialWaveFunction>& wf_list,
```
Comment (ye-luo) [inline src/QMCWaveFunctions/TrialWaveFunction.h:322]: Please make this static like other flex_ functions.

File: src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp [hunk 3]
```diff
@@ -904,6 +904,13 @@ void LCAOrbitalBuilder::EvalPeriodicImagePhaseFactors(PosType SuperTwist,
 void LCAOrbitalBuilder::EvalPeriodicImagePhaseFactors(PosType SuperTwist,
                                                       std::vector<std::complex<RealType>>& LocPeriodicImagePhaseFactors)
 {
+  // For open boundary conditions
+  if (PBCImages[0] == 0 && PBCImages[1] == 0 && PBCImages[2] == 0) {
+    LocPeriodicImagePhaseFactors.resize(1);
+    LocPeriodicImagePhaseFactors[0] = 1.0;
+    return;
+  }
+
```
Comment (ye-luo) [inline src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp:913]: I feel better not using early return.
How about protect hdf5 read with
```
if  (!(PBCImages[0] == 0 && PBCImages[1] == 0 && PBCImages[2] == 0))
```

File: src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp [hunk 4]
```diff
@@ -358,4 +359,167 @@ TEST_CASE("TrialWaveFunction_diamondC_1x1x1", "[wavefunction]")
 
 }
 
+void setup_He_wavefunction(Communicate* c, ParticleSet** elec_ptr, ParticleSet** ions_ptr, TrialWaveFunction** psi)
```
Comment (ye-luo) [inline src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp:362]: How about move this test to a separate file?
```
test_TrialWaveFunction_He.cpp
```

File: src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp [hunk 5]
```diff
@@ -358,4 +359,167 @@ TEST_CASE("TrialWaveFunction_diamondC_1x1x1", "[wavefunction]")
...
+{
+  ParticleSet* elec = new ParticleSet;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp:364]: Can we avoid using pointers in this file?

File: src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp (hunk 3 above)
Comment (markdewing) [reply src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp:913]: I'm concerned that avoiding the early return is going to complicate the logic and make it harder to read.

File: src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp (hunk 5 above)
Comment (markdewing) [reply src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp]: Fixed.    Used std::unique_ptr for WaveFunctionFactory, so at least now it will get deleted when the test ends.

File: src/Containers/MinimalContainers/RecordArray.hpp (hunk 1 above)
Comment (markdewing) [reply src/Containers/MinimalContainers/RecordArray.hpp]: removed

File: src/QMCWaveFunctions/TrialWaveFunction.h (hunk 2 above)
Comment (markdewing) [reply src/QMCWaveFunctions/TrialWaveFunction.h]: Fixed.

File: src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp (hunk 3 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp:913]: To me adding early return creates two fully divergent code paths and complicates the logic. If we don't need to distinguish no-PBC and PBC, it is better not to do it. However, we need to add proper comment saying PBC case needs to read lattice vectors.

Comment (ye-luo) [reply src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp:913]: I will leave it to others to express their preferences.

Comment (markdewing) [reply src/QMCWaveFunctions/lcao/LCAOrbitalBuilder.cpp:913]: I worked through the TransX logic in the loop below the HDF file read and verified it gives the same result, and I found a way to add the extra logic.   I will update with the early return removed as soon as the local build works.

**Comment by prckent** [comment]
@anbenali Comments?
@markdewing Make an issue for Anouar if we don't hear. Lets merge and continue.
